src/instructlab/model/download.py

```diff
@@ -12,6 +16,232 @@
 # First Party
 from instructlab import clickext
 from instructlab.configuration import DEFAULTS
+from instructlab.utils import is_oci_repo, is_tool_present
+
+
+class Downloader(abc.ABC):
+    """Base class for a downloading backend"""
+
+    def __init__(
+        self,
+        repository: str,
+        release: str,
+        download_dest: str,
+    ) -> None:
+        self.repository = repository
+        self.release = release
+        self.download_dest = download_dest
+
+    @abc.abstractmethod
+    def download(self) -> None:
+        """Downloads model from specified repo/release and stores it into download_dest"""
+
+
+class HFDownloader(Downloader):
+    """Class to handle downloading safetensors and GGUF models from Huggingface"""
+
+    def __init__(
+        self,
+        repository: str,
+        release: str,
+        download_dest: str,
+        filename: str,
+        hf_token: str,
+        ctx,
+    ) -> None:
+        super().__init__(
+            repository=repository, release=release, download_dest=download_dest
+        )
+        self.repository = repository
+        self.release = release
+        self.download_dest = download_dest
+        self.filename = filename
+        self.hf_token = hf_token
+        self.ctx = ctx
+
+    def download(self):
+        """Download the model(s) to train"""
+        click.echo(
+            f"Downloading model from huggingface: {self.repository}@{self.release} to {self.download_dest}..."
+        )
+
+        if self.hf_token == "" and "instructlab" not in self.repository:
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to download HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
+
+        try:
+            if self.ctx.obj is not None:
+                hf_logging.set_verbosity(self.ctx.obj.config.general.log_level.upper())
+            files = list_repo_files(repo_id=self.repository, token=self.hf_token)
+            if any(".safetensors" in string for string in files):
+                self.download_safetensors()
+            else:
+                self.download_gguf()
+
+        except Exception as exc:
```

Comment:
what kind of exceptions are there here? I am fine with this for now if its too hard to catch all different types

Reply:
unsure, I don't see `hf_hub_download` raising many exceptions 
https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/file_download.py#L1010

Reply:
This code just moved so the line was already present.